File: Dataset/Leetcode/valid/62/317.java
 class Solution {
    public int XXX(int m, int n) {
        int[][] dp = new int[m][n];
        for(int i = 0; i < m; i++){dp[i][0] = 1;}// 第0列全部都是1
        for(int i = 0; i < n; i++){dp[0][i] = 1;}// 第0行全部都是1
        for(int i = 1; i < m; i++){
            for(int j = 1; j < n; j++){
                dp[i][j] = dp[i-1][j]+dp[i][j-1];//从上方下来和从左方过来
            }
        }
        return dp[m-1][n-1];
    }
}

File: Dataset/Leetcode/train/11/317.py
class Solution:
    def XXX(self, height: List[int]) -> int:
        # 此方法超时
        # max_area = 0
        # for i, h in enumerate(height):
        #     for j in range(i+1, len(height)):
        #         tmp_area = (j-i) * min(height[i], height[j])
        #         max_area = max(max_area, tmp_area)
        # return max_area
        
        # 双指针算法
        left = 0
        right = len(height) - 1
        max_area = 0
        while left < right:
            tmp_area = min(height[right], height[left]) * (right - left)
            max_area = max(tmp_area, max_area)

            # 比较左右边界，高度较小的往中间移动
            if height[left] < height[right]:
                left += 1
            else:
                right -= 1
        return max_area


